test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,974 @@
+
+  stream.on('data', message => {
+    if (message.msg.includes('FAIL ON LISTEN')) {
+      t.pass('Logged Error Message')
+    }
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function () {
+    t.pass('called onListen error')
+    throw new Error('FAIL ON LISTEN')
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('onListen hooks continue after error')
+    done()
+  })
+
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+
+t.test('localhost onListen async should log errors as warnings and continue', async t => {
+  t.plan(4)
+  const stream = split(JSON.parse)
+  const fastify = Fastify({
+    forceCloseConnections: false,
+    logger: {
+      stream,
+      level: 'info'
+    }
+  })
+  t.teardown(fastify.close.bind(fastify))
+
+  stream.on('data', message => {
+    if (message.msg.includes('FAIL ON LISTEN')) {
+      t.pass('Logged Error Message')
+    }
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.pass('called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.pass('called onListen error')
+    throw new Error('FAIL ON LISTEN')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.pass('onListen hooks continue after error')
+  })
+
+  await fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+
+t.test('localhost Register onListen hook after a plugin inside a plugin', t => {
+  t.plan(3)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function (done) {
+      t.pass('called')
+      done()
+    })
+    done()
+  }))
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function (done) {
+      t.pass('called')
+      done()
+    })
+
+    instance.addHook('onListen', function (done) {
+      t.pass('called')
+      done()
+    })
+
+    done()
+  }))
+
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+
+t.test('localhost Register onListen hook after a plugin inside a plugin should log errors as warnings and continue', t => {
+  t.plan(6)
+  const stream = split(JSON.parse)
+  const fastify = Fastify({
+    forceCloseConnections: false,
+    logger: {
+      stream,
+      level: 'info'
+    }
+  })
+  t.teardown(fastify.close.bind(fastify))
+
+  stream.on('data', message => {
+    if (message.msg.includes('Plugin Error')) {
+      t.pass('Logged Error Message')
+    }
+  })
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function () {
+      t.pass('called')
+      throw new Error('Plugin Error')
+    })
+    done()
+  }))
+
+  fastify.register(fp(function (instance, opts, done) {
+    instance.addHook('onListen', function () {
+      t.pass('called')
+      throw new Error('Plugin Error')
+    })
```

Comment:
```suggestion
    instance.addHook('onListen', function (done) {
      t.pass('called')
      done(new Error('Plugin Error'))
    })
```

Reply:
Due to our current implementation making our functions async, should we keep the test the way it is, or would it be expected to call the callback?

Reply:
There are basically two cases. The function itself throws an error, or we pass an Error to done. In both cases the error gets logged now.